src/LinearRemapFV.cpp

```diff
@@ -1611,199 +1610,451 @@ void LinearRemapGeneralizedBarycentric(
+				
+				nodesDualFace.push_back(meshInputDual.nodes[faceDualFace[i]]);
+				
+			}
+			
+			//Vector of contributing weights
+			
+			std::vector<double> vecWeights(iEdges);
+			
+			//Compute generalized barycentric coordinates
+			
+			GeneralizedBarycentricCoordinates(nodeQ, nodesDualFace, vecWeights);	
+		
+			//Add weights to map
+			
+			for (int i = 0; i < iEdges; i++){
 
-						vecTriangleSubAreas[m] = dSubAreaM;
+				int iContributingFaceI = meshInputDual.faces[iFaceFinal][i];
 
-						double dTriangleAreaM = CalculateFaceArea(faceTriangleM, nodesTriangleM);
+				smatMap(ixFirst, iContributingFaceI) = vecWeights[i];
+				
+			}
+			
+		}
+		
+	}
+	
+	else{
+
+		//kd-tree of dual mesh centers
+		kdtree * kdTarget = kd_create(3);
+    
+		//Vector of source face centers
+		NodeVector vecSourceCenterArray;
+    
+		//Vector of target face centers
+		NodeVector vecTargetCenterArray;
+
+		// Vector of centers of the source mesh
+		for (int i = 0; i < meshInput.faces.size(); i++){
+	
+			const Face & face = meshInput.faces[i];
+	
+			Node nodeCentroid = GetFaceCentroid(face, meshInput.nodes);
+			
+			nodeCentroid = nodeCentroid.Normalized();
+			
+			vecSourceCenterArray.push_back(nodeCentroid);
+	
+				
+		}
+		
+		for (int i = 0; i < meshOutput.faces.size(); i++){
 
-						vecTriangleAreas[m] = dTriangleAreaM;
+			const Face & face = meshOutput.faces[i];
 
+			Node nodeCentroid = GetFaceCentroid(face, meshOutput.nodes);
+		
+			nodeCentroid = nodeCentroid.Normalized();
 
+			kd_insert3(
+				kdTarget,
+				nodeCentroid.x,
+				nodeCentroid.y,
+				nodeCentroid.z,
+				(void*)(&(meshOutput.faces[i])));
+			
+			vecTargetCenterArray.push_back(nodeCentroid);
+			
+		}
+		
+		std::set<int> setAllTargets;
+		
+		// Loop through all source faces
+		
+		for (int ixFirst = 0; ixFirst < meshInput.faces.size(); ixFirst++) {
+	
+			// Output every 1000 source elements
+			if (ixFirst % 1000 == 0) {
+				
+				Announce("Element %i/%i", ixFirst, meshInput.faces.size());
+				
+			}
+			
+			// This Face
+			const Face & faceFirst = meshInput.faces[ixFirst];
+			
+			//Find the maximum distance from the current source face center to its nodes
+			
+			double dMaxNodeDist = 0.0;
+			
+			Node faceFirstCenter = vecSourceCenterArray[ixFirst];
+			
+			int iEdges = faceFirst.edges.size();
+			
+			for (int i = 0; i < iEdges; i++){
+				
+				//Great circle distance is just the arccos of dot product
+				
+				Node nodeI = meshInput.nodes[faceFirst[i]];
+				
+				double dDistToNodeI = acos(DotProduct(nodeI, faceFirstCenter));
+				
+				if ( dDistToNodeI > dMaxNodeDist ){
+					
+					dMaxNodeDist = dDistToNodeI;
+					
+				}
+				
+			}
+			
+			//Find all the target face centers that are within a given distance of the given source face center
+			
+			/* find points closest to the origin and within distance radius */
+			//struct *presults;
+										 
+			struct kdres *presults = kd_nearest_range3(kdTarget,faceFirstCenter.x, faceFirstCenter.y, 
```

Comment:
You have to be somewhat careful for distance-based nearest neighbor queries to make sure it doesn't return points across a land boundary etc when there is a thin peninsula for example.

Reply:
Here, the distance-based nearest neighbor is used only to find a set of target faces whose centers could potentially be within a given source face.  Once we have that set of target faces, we then check each face to see whether its center is contained within that source face.  If the center is not contained within the given source face, it will be ignored.